libs/openFrameworks/utils/ofThreadChannel.h

```diff
@@ -256,6 +256,153 @@ class ofThreadChannel{
 		return true;
 	}
 
+	/// \brief Parameter used in trySend method.
+	///
+	/// This parameter determines what strategy to follow when queue size > maxSize
+	/// argument when we call trySend
+	enum TrySendStrategy {
+		/// Do not discard, i.e. same as send(...)
+		None = 0,
+
+		/// If the queue size is >= maxSize argument, discard this message
+		DiscardLast,
+
+		/// If the queue size is >= maxSize argument, discard oldest message and add this message
+		DiscardFirst,
+
+		/// Wait until queue size is < maxSize and then push the message onto the queue
+		WaitForSpace
+	};
+
+	/// \brief Message returned from trySend method
+	struct TrySendResult {
+		bool queueTrimmed;
+		bool messageSent;
+		bool channelOpen;
+	};
+
+	/// \brief Send a value to the receiver by making a copy and apply the DiscardStrategy if
+	/// the queue size would become > maxSize.
+	///
+	/// This method copies the contents of the sent value, leaving the original
+	/// data unchanged.
+	///
+	/// \returns TrySendResult
+	TrySendResult trySend(T & value, const DiscardStrategy & discardStrategy, size_t maxSize) {
+		std::unique_lock<std::mutex> lock(mutex);
+		if (closed) {
+			return TrySendResult({
+				false,
+				false,
+				false
+			});
+		}
+		if (queue.size() >= maxSize) {
+			switch (discardStrategy) {
+			case DiscardStrategy::DiscardLast:
+			{
+				return TrySendResult({
+					false,
+					false,
+					true
+				});
+			}
+			case DiscardStrategy::DiscardFirst:
+			{
+				do {
+					this->queue.pop();
+				} while (queue.size() >= maxSize)
+				queue.push(value);
+				condition.notify_all();
+				return TrySendResult({
+					true,
+					true,
+					true
+				});
+			}
+			case DiscardStrategy::WaitForSpace:
+			{
+				do {
+					lock.unlock();
+					lock.lock();
+				} while (queue.size() >= maxSize)
```

Comment:
This do while is totally evil
but what is the best strategy?
Ideal is wait until next release of lock (e.g. keep lock unlocked, and wait until next unlock).

We could add another conditional value which `receive` but that seems 'selfish' for this function to add weight to receive. Or we could add an optional sleep time argument, but sleep is always at least 50ms AFAIK.